src/arch/xtensa/gdb/debugexception.S

```diff
@@ -0,0 +1,235 @@
+ *   * Neither the name of the Intel Corporation nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Marcin Rajwa <marcin.rajwa@linux.intel.com>
+ *
+ * Debug context switch.
+ *
+ */
+
+#include <arch/gdb/xtensa-defs.h>
+#define BREAK_INS_IDENTIFIER 0x40
+#define BREAK_INS_SIZE 0x03
+#define PS_EXCM_MASK 0xFFFFFFEF
+#define PS_EXCM_EXCEPTION_MODE_ON 0x10
+#define DEBUG_GDB_MEM_LOC 0x9E008060
+
+/*
+Save special register designated by 'reg' into
+backup space calculated by offset 'loc * 4' from
+memory pointed by a3.
+*/
+.macro	SAVE_	reg, loc
+	rsr	a1, \reg
+	s32i	a1, a3, \loc * 4
+.endm
+
+.macro	SAVE	reg
+	SAVE_	\reg, \reg
+.endm
+
+/*
+Load special register designated by 'reg' from
+backup space calculated by offset 'loc * 4' from
+memory pointed by a3.
+*/
+.macro	LOAD_	reg, loc
+	l32i	a1, a3, \loc * 4
+	wsr	a1, \reg
+.endm
+
+.macro	LOAD	reg
+	LOAD_	\reg, \reg
+.endm
+
+
+
+.text
+
+/*
+Temporary stack for context switch
+TODO: move it to dedicated GDB_STACK section
+*/
+
+gdb_stack:
+	.fill 0x1000 , 4 , 0
+gdb_stack_end:
+
+.global DebugExceptionEntry
+.align	4
+
+/*
+Backup important special registers plus
+all regular ones (whole register file).
+Change EXCM field of PS back to normal mode
+after an interrupt took place.
+*/
+DebugExceptionEntry:
+
+	movi	a2, aregs
+	s32i	a0, a2, 0
+	s32i	a1, a2, 4
+	rsr	a1, DEBUG_EXCSAVE
+	s32i	a1, a2, 8
+	s32i	a3, a2, 12
+
+	movi	a3, sregs
+	SAVE	LBEG
+	SAVE	LEND
+	SAVE	LCOUNT
+	SAVE	SAR
+	SAVE	WINDOWBASE
+	SAVE	WINDOWSTART
+
+	rsr	a1, DEBUG_PC
+	l8ui	a2, a1, 1
+	movi 	a0, BREAK_INS_IDENTIFIER
+	bne	a2, a0, 1f
+	addi	a1, a1, BREAK_INS_SIZE
+1:
+	s32i	a1, a3, DEBUG_PC * 4
+
+	SAVE	EXCSAVE_1
+	SAVE_	DEBUG_PS, PS
+	SAVE	EXCCAUSE
+	SAVE	DEBUGCAUSE
+	SAVE	EXCVADDR
+
+	movi	a1, XCHAL_NUM_AREGS / 4 - 1
+	movi	a2, aregs
+1:
+	s32i	a4, a2, 16
+	s32i	a5, a2, 20
+	s32i	a6, a2, 24
+	s32i	a7, a2, 28
+
+	addi	a6, a2, 16
+	addi	a5, a1, -1
+	rotw	1
+	bnez	a1, 1b
+
+	movi	a1, 1
+	wsr	a1, windowstart
+	movi	a0, 0
+	wsr	a0, windowbase
+	rsync
+
+	movi	a0, 0
+	movi	a1, gdb_stack + STACK_SIZE - 20
```

Comment:
likewise, best to comment why we subtract 20 from stack size, can this be a macro or calculated by preprocessor ?

Reply:
Done.